waspc/cli/src/Wasp/Cli/GithubRepo.hs

```diff
@@ -0,0 +1,129 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Wasp.Cli.GithubRepo where
+
+import qualified Codec.Archive.Tar as Tar
+import qualified Codec.Compression.GZip as GZip
+import Control.Exception (SomeException, try)
+import Data.Aeson
+  ( FromJSON,
+    parseJSON,
+    withObject,
+    (.:),
+  )
+import qualified Data.ByteString.Lazy as BL
+import Data.Functor ((<&>))
+import Data.List (intercalate)
+import Data.Maybe (fromJust, maybeToList)
+import Network.HTTP.Conduit (simpleHttp)
+import qualified Network.HTTP.Simple as HTTP
+import Path.IO (copyDirRecur, removeDirRecur)
+import StrongPath (Abs, Dir, File, Path', (</>))
+import qualified StrongPath as SP
+import StrongPath.Path (toPathAbsDir)
+import System.Directory (createDirectoryIfMissing)
+import Wasp.Cli.FileSystem (getUserCacheDir)
+
+type GithubRepo = (GithubRepoOwner, GithubRepoName)
+
+type GithubRepoOwner = String
+
+type GithubRepoName = String
+
+fetchFolderFromGithubRepoToDisk ::
+  forall d.
+  GithubRepo ->
+  String ->
+  Path' Abs (Dir d) ->
+  IO (Either String ())
+fetchFolderFromGithubRepoToDisk (githubOrg, repoName) pathToFolderInRepo destinationOnDisk = do
```

Comment:
This whole function, well its helper functions in `where`, are a bit convoluted I think -> there are some assumptions going around, especially around the archive, what it looks like, ... . I commented on some specific things but there might be some more that I just couldn't pinpoint yet, so I suggest trying to look into it very carefully, looking for those assumptions, trying to make every function as standalone, readable and simple as possible, while being explicit about what it depends on, what it outputs, ... . So that data flow is clear.

Reply:
I've rewritten the file quite a bit to make it more simple and all the functions are more agnostic.

Reply:
We went through this offline and it should be much better